### DISCUSSION TITLE
How to correctly deploy your AAB to your device

### DISCUSSION THREAD

**Discussion by stoff99** [airsdk/Adobe-Runtime-Support#994] [General]
As you can read in the AIR release not PDF there is the documentation to extract the akps out of the aab and deploy it on the connected device.
1. bundletool build-apks --bundle YourAPP.aab --output output.apks --connected-device
2. bundletool install-apks --apks=output.apks

This is the fastest way to test your aab release on a real device.
But i think i miss some parameter in this commands because i miss some resources from ANEs.

I'm using the Distriqt Firebase and Push Notification ANEs for Firebase Cloud Messaging. You need to package the firebase configuration (values.xml) inside of the custom resources ANE. I think the most of you know this ANEs and the process.
This values.xml is missing in the exported APKS file.

But when i deploy the app via Google Play Store beta testing, everything is working fine.
So there is some difference between the bundletool and google play providing the apks file.

How do you deploy your test builds when you want to quickly test your app? 
Do you know the missing parameter for the bundletool?

Thx

**Comment by ajwfrost** [comment]
Hi

These files are all actually just zip files, so I would suggest you open each of them in turn to see whether you can track where the configuration file is going... if it's called 'values.xml' then I'm a bit concerned that it may be being merged with something else, or overwritten perhaps! or maybe the bundletool doesn't think it's referenced so isn't putting it into the APK collection..

So:
- try unzipping the .aab file and checking the resulting folder structure to find the values.xml file and ensuring your values are inside it
- then generate the .apks file and unzip that, within which are a bunch of APK files. Each of these can also be unzipped to see whether the values.xml file is in there
- it might be that you can't actually see the values.xml if it's compiled into the resources, so you may need to do the "Analyze APK/AAB" option from Android Studio.

Let us know how you get on; it may also be worth ensuring you've got the latest version of bundletool from Google, in case this is just a bug: https://github.com/google/bundletool/releases

thanks

**Reply by stoff99** [reply]
Hey @ajwfrost ,
i did the deep check already and can't find the file or the content of it.
You can see my investigation here: https://github.com/distriqt/ANE-PushNotifications/issues/497#issuecomment-882262282

I have the latest bundletool installed (1.7.0)

**Reply by ajwfrost** [reply]
Okay thanks - so with the APK file there, you can see that ADT just embeds the whole of the extension's Android-ARM64 subfolder into this location, with the custom resources added in there; they're not actually used from there though, as part of the packaging process should mean that they are built into the main application resources and the R.xx.xx value is created at that point...

With an App Bundle though, we put each set of resources into a named package that is built as part of the overall application - so the files wouldn't be sat in the same place, they would be compiled into the resources that's then - hopefully - exported as part of the Bundletool action.

So I suspect something may be going wrong due to the package naming, maybe a conflict or a file overwrite due to the use of the CustomResources ANE which we perhaps aren't handling properly.

Are you able to send us your custom resources ANE file, along with your main application descriptor file, and we can see if we can reproduce this here...
https://transfer.harman.com/requests/BjdKlY9medu72QoaN4sP5D for upload..

thanks

**Reply by stoff99** [reply]
Hey, i sent you the files.
Thx

**Reply by ajwfrost** [reply]
Thanks. So, I think we're hitting a limitation of the Android Studio tools here. Within an APK, the resources you provide in the CustomResources ANE are included under a different package name so you can access them via [your-custome-resourcepackage-name].R.strings.some_definition. This isn't actually reflected in what we see when analysing an APK, even one generated directly...  which means we can't confirm the suspicions here.

How are you setting the values up where you use them? or is this something that's handled internally within the Firebase ANE? @marchbold tagging you to help us understand the Android/Java side of things here: if there are strings such as `google_api_key` defined via a custom resource ANE, is this what you're then using to populate the Firebase configuration? and do you then make an assumption that these are available in a particular package name? Or, do you expose these to ActionScript and have this set up the Firebase configuration?

I'm not sure whether this is a bug (or limitation?) in bundletool or whether there's some packaging that's going wrong; the app bundle you said worked okay?

I did notice a reason code `API_KEY_ANDROID_APP_BLOCKED` which perhaps means that the key wasn't provided properly, or could it be that they've got your key but have blocked it..?

thanks

**Reply by stoff99** [reply]
@ajwfrost exactly, for the distriqt firebase and push notification using firebase cloud messaging you have to provide this customeResources ANE which includes the push icons and the firebase configuration (values.xml).
The ANEs handle this internally than.

When i check logcat, installing and starting the app with bundletool i see this error message:
![logcat_log_und__Android__Firebase_messaging_isn_t_working_anymore_when_deploy_app_as_aab_·_Issue__497_·_distriqt_ANE-PushNotifications](https://user-images.githubusercontent.com/892820/126483495-f212f65d-2955-494f-8540-0a5612a6248c.jpg)

So @marchbold and i found out it doesn't get the configuration from the resources ANE.

**Reply by stoff99** [reply]
Ho do you actually test your aab releases? Only via Google Play internal testing functionality?

**Reply by ajwfrost** [reply]
Actually we test the AABs by turning them into APKs which is why I'm curious as to the mechanism that it uses to pick up the value. The string resource appears to be present, and we had already checked to ensure that these embedded resources are being compiled properly - so it's then a case of how the Firebase ANE is loading the value. But it definitely seems odd that it works if the APK provided from the Play Store is working for you whereas the APK provided from the bundletool is not...

I'd be keen to see a log that includes the value that the Firebase ANE has retrieved from the resources bundle and is setting on that configuration... we can perhaps try to replicate the same set-up here, we've got ANEs loading resources from their own packages but I don't know that we have a test case that covers one ANE that depends on the contents of another..

thanks

**Reply by marchbold** [reply]
Firstly, these particular resources are configuration values provided to firebase / google services and should be in the applications package's `R` class.

This sounds related to the discussion we were having around the Google Play Services Base extension. I believe the way the "package name" for a resources directory was used in the old packaging method isn't what we think. It never seemed to match up with what we expected for sure. 

The custom resources ANE is one in particular. Initially we had set this up with the ability to build a custom android project with a specified package name and then pass that into the resources, however we eventually found all that was a waste of time. You could provide any package name and the resources would be found. There was something odd going on. However hopefully with the new gradle method, once this is applied to both packaging methods we can forget about all the time I've wasted on resources......

So most people using our custom resources ANE probably have a package name of `com.distriqt.extension.customresources` for the package name and this works, ie Firebase can find these resources that are supposed to be in the application package `.R` class. 

However if you are actually applying this correctly now then these resources are probably getting lost.

What really stumps me is why they are being found when it is distributed via the play store?

**Reply by marchbold** [reply]
I would like to suggest @stoff99 try using the `resdir` approach, i.e. not using a native extension

So remove the Custom Resources ANE and create a directory in your source tree, something like this:

```
res
|____values
| |____values.xml
|____drawable-xhdpi
| |____ic_stat_distriqt_default.png
|____drawable-xxhdpi
| |____ic_stat_distriqt_default.png
|____drawable-hdpi
| |____ic_stat_distriqt_default.png
|____drawable-mdpi
| |____ic_stat_distriqt_default.png
|____drawable-xxxhdpi
| |____ic_stat_distriqt_default.png
```

Then add `<resdir>res</resdir>` to your application descriptor just after the android node:

```xml

        ]]></manifestAdditions>
    </android>
    <resdir>res</resdir>

```

We have had some user's report this is failing (unsure of specifics yet) but I believe this should be how we move forward with resources like this..

**Reply by stoff99** [reply]
Hey, i will try that as soon as possible. 👍

**Reply by ajwfrost** [reply]
Thanks both.

@marchbold I'm assuming that when you create the .jar file for the Firebase ANE, there would be some dummy values.xml in there somewhere so that it can compile with an `R.strings.something` format? but if it's just stated like that, then wouldn't the `R` belong to the Firebase package (com.distriqt namespace)?

Interestingly if you look at the compiled .dex file in an APK generated in the 'old' way, you can see the `R$string.google_api_key` resource value - and the AIR-internal ones like `IDA_APP_WAITING_DEBUGGER_WARNING` - under `com.distriqt.extenion.customResources` as well as under the application's own `air.com.harman.air.TestPackage` namespace...  which is odd, but perhaps why it worked before.

If I inspect the AAB file and the .dex there, I see the `R$string.google_api_key` under `com.distriqt.extenion.customResources` but not the AIR-internal ones ... but then under our application ID I see both of them! Which is very odd. The .dex file in a generated APK (from bundletool) shows the same. But part of me is wondering whether it's actually correct...

If you then look at an APK generated from bundletool, the resources.arsc file has only our own application package ID listed, and the google_apk_key value here is just counted under the rest of the built-in AIR resources, so it would definitely not have the same value as the compiled code from your .jar file... maybe there is a limitation of the resource viewer in Android Studio but it doesn't give me the option to select another package name. Or perhaps that's the problem, it hasn't included the other package name in there! Sadly Android Studio isn't showing anything when I try to look at the resource.pb component from the App Bundle itself...

It seems a bit of a dark art, trying to get resources to be shared between different components, but it would be nice if we could somehow adapt the build process to ensure that it works with existing ANEs and with both APKs built via ADT as well as via AAB and bundletool... so we can investigate a bit further..

**Reply by marchbold** [reply]
Just to clarify, these aren't resources that we access through an extension package name. They are used by the Google / Firebase libs internally and are expected to be in application package `R` class. They are a bit unique in this way.  The "Custom Resources" ANE was designed to be used to add resources to the application package namespace. 

That bit about resources being listed twice, would explain why these worked. i.e. if AIR listed them under the package name we provided and also under the application package name then the later is where Firebase and our extensions are looking for these values. 

I wouldn't suggest trying to replicate an old bug though! I would prefer devs to move away from the Custom Resources ANE now that we have the resdir facilities. It was only ever created to resolve this missing functionality.

**Reply by ajwfrost** [reply]
Okay thanks for the background, I hadn't realised they were accessed internally by Firebase!
I think, in case you wanted to still use the 'custom resources' ANE, it might be that this could be set up with the resource package as the same as the application package ID but I suspect the ADT logic would break down at that point and it would assume this is a new namespace -> creating a new folder (rather than using 'app') and probably confusing gradle! Perhaps we need to handle that as a generic case anyway, but it may even be simpler for us to check specifically for com.distriqt.extension.customResources and handle it as you had expected/as it was originally.

(Side note: I'm aware of a few low-level "bugs" in the Flash Player/AIR runtime that I'd discussed with Adobe and they had said not to fix them as people depend on the "incorrect" functionality ... I can see what they mean now!)

Anyway -> with the Android Studio project set-up, we should be able to test some of these scenarios to see how we're able to access the resources in the package and maybe see how to get this working ..!

thanks

**Reply by marchbold** [reply]
ha ha yeah, I totally understand that...  :) 

I honestly would prefer to deprecate the Custom Resources ANE and move people to using `resdir` and I think this is a good time with the aab introduction.  Most people using it, simply have `values.xml` in there with the firebase config and it's such overkill for that, and it will be much easier for them to view / modify it without the ANE I doubt anyone will object to it's removal.

**Reply by stoff99** [reply]
@marchbold totally true! ;)

**Reply by leossmith** [reply]
Looks like this is also related to #934 latest issue. Works with customresource doesnt work with resdir. Indeed would be convenient to not having to rely on an ANE for just values.xml and maybe some icons.

**Reply by marchbold** [reply]
@leosmith I was going to loop you in here, did you look any further into why it didn't work with resdir? did you keep the logs at all?

**Reply by leossmith** [reply]
I am not actually getting any errors other than the one i posted in the issue. App just hangs and never starts

![image](https://user-images.githubusercontent.com/4300876/126642017-a408f843-5cc7-4670-8495-09869e8f00fb.png)

**Reply by ajwfrost** [reply]
Hi @leossmith  - on the resdir thing, what files are in your folder there? I'm wondering whether there's a conflict.... just trying a few things here and I get build errors with most of what I try e.g. providing "strings.xml" under my `resfolder/values` location, then it overrides the strings.xml that was pulled in from the AIR SDK, and the build then fails due to missing application names etc. But possibly there is an override that has not been picked up at compile time, but is causing a failure at runtime...

**Reply by marchbold** [reply]
Most likely there will be a `values/values.xml` as this is what we generate for the firebase config and then often there are some notification icons, like what I listed above.

**Reply by leossmith** [reply]
There is a values/values.xml, drawable folders for adaptive icons and drawable folders for push notification assets (icons).

**Reply by stoff99** [reply]
@marchbold i tried the solution with the <resdir>res</resdir>, i removed the customResources ane but its still not working.
Same result in the logcat as before.

**Reply by marchbold** [reply]
@stoff99 Ah that's a pity, could you enable Firebase debug outputs and post the logs again? 

https://docs.airnativeextensions.com/docs/firebase/core/analytics#android

Also just confirming, did this fail when installed through Google Play or just through the bundletool ?

**Reply by ajwfrost** [reply]
Also if it would be possible for us to get a copy of your generated AAB project, it would help us to see what might be going wrong ... if you change the lib/adt.cfg file so that it has `KeepAndroidStudioOutput=true` and then zip up the generated `AndroidStudioProject` folder (you can delete the 'build' subfolders to keep the size down)...

Upload link: https://transfer.harman.com/requests/7icDH4tqiQCdKYi46U1uN0

thanks

**Reply by stoff99** [reply]
Hey @marchbold i made a mistake. I forget to comment out <extensionID>com.distriqt.CustomResources</extensionID> in the app xml file.
As i commented out i get following compiler error:
![Error_Message_und_ETeamApp_–_ETeamApp-app_xml__ETeamApp_](https://user-images.githubusercontent.com/892820/127620876-f20598c8-6f65-4b26-b0af-1c5eac931d81.jpg)

i have the res dir added like this:
![ETeamApp_–_ETeamApp-app_xml__ETeamApp_](https://user-images.githubusercontent.com/892820/127621003-11d27579-0d10-4149-bfb1-36e094485efb.jpg)

With the CustomResources its working when i deploy via PlayStore.
I'm doing something wrong?

**Reply by marchbold** [reply]
Are you including it in your module, ie has it been copied to the out directory?

**Reply by stoff99** [reply]
@ajwfrost i sent you the zip file.
@marchbold i copied the res file now manually in the out folder.
Now it compiles correctly. Currently i upload to play store to test

**Reply by stoff99** [reply]
@marchbold sry for my fail, with copy the res manually to the out folder its working now.
i also receive pushes and see no error in logcat when i deploy it with play store.
But when i deploy it with bundletool i still have the same error.


@ajwfrost: If you want, send me another upload link, so i will upload the latest AndroidStudio output with the res folder solution.

**Reply by ajwfrost** [reply]
@stoff99 are you manually copying the res folder? as this looks like it didn't happen in the package you sent me, but the application descriptor file doesn't have the `resdir` entry in it which may be why? unless you can add it to the command line like `-resdir res` instead? See Michael's suggestion from 8 days ago:
> Then add `<resdir>res</resdir>` to your application descriptor just after the android node

From the log, the firebase stuff is happening in another thread. I'm wondering whether it's a multidex issue. If you open the generated project folder in Android Studio and run/debug from there, it would create an APK file -> does that one work? I'm still a little uncertain as to how the Firebase library is getting the configuration values it needs..

Could you perhaps use the bundletool to create an APKS set for the connected device, and send us that plus the APK file from the above test, we then can look at the differences...

https://transfer.harman.com/requests/VLzovqj1gZxYbNyxNRLGR8

thanks

**Reply by stoff99** [reply]
Hey @ajwfrost,
read my last comment please. I was able to build it now correctly with the res folder solution.
Please check again my Android Studio zip which i uploaded to your lastest link. I also added the aab file. https://transfer.harman.com/requests/VLzovqj1gZxYbNyxNRLGR8

And just again for notice:
Deploy via PlayStore internal testing =  Everything is working
Deploy with bundletool = not working

**Reply by ajwfrost** [reply]
Hi @stoff99 - so from what we can see, it is possible for the extensions to pick up the resource values from your values.xml file regardless of how the multidex feature is working (and actually the R$string classes all seem to be in the first classes.dex other than in a direct APK build!)..

And actually looking again at the log file you provided, this does appear to be the case:
```
07-30 10:58:19.218  9412  9446 W Firebase-Installations: Error when communicating with the Firebase Installations server API.
07-30 10:58:19.218  9412  9446 W Firebase-Installations: Firebase options used while communicating with Firebase server APIs: <<<google_api_key>>>, <<<project_id>>>, <<<google_app_id>>>
07-30 10:58:19.218  9412  9446 E Firebase-Installations: Firebase Installations can not communicate with Firebase server APIs due to invalid configuration. Please update your Firebase initialization process and set valid Firebase options (API key, Project ID, Application ID) when initializing Firebase.
```

I've changed the actual values from the log so that you can see what elements from within your `values.xml` file these are taken from... but look at the log and you'll see that it's correctly picked up the values from your xml.

So, the Firebase request appears to be using the right values, but the connection is getting rejected. The error message is that `Requests from this Android client application air.com.sf.eteam.dev2 are blocked` so I'm wondering how your Firebase permissions are set up, and I'm also wondering whether this initial "air." value is causing some confusion.

Could you perhaps try rebuilding with the lib\adt.cfg setting `AddAirToAppID=false` so that we don't prepend your package name with this?

It doesn't help explain why there's a difference between the PlayStore deployment and the local one though...

thanks

**Reply by stoff99** [reply]
Hello @ajwfrost ,

you are completely right, i don't know why i did't recognize that the log already trace out the project id and so on ......
In this case, for sure, it can read the values ... :D sry for that.

I tried now to build the app without the air. I also the app id in the ANE services and providers in the android app xml.
But the result is the same, see the attached logcat log:
[LogWithoutAirInAppId.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6914844/LogWithoutAirInAppId.txt)

Looks like there is some difference for the OAuth request to firebase between local and store deployment.

**Reply by ajwfrost** [reply]
For the store deployment, presumably you are using the Google Play App Signing? I don't know fully how this works but am wondering whether it's working when the request (or app?) is signed with the key from the play store, but not when it's signed with your 'local' or upload key?

**Reply by stoff99** [reply]
yes that could happen, i use the google play signing because its required for aab deployment.
So for me its not a blocker that the local deployment isn't working now correctly but it would just be fin. :)

**Reply by GeneralVimes** [reply]
> These files are all actually just zip files

Hello, Andrew!
Looks like today we together with @lpmitchell found a case when this .zip file becomes corrupted.

Just today I built the bundle with the newest AIR. I even wrote a how-to for other developers: https://www.reddit.com/r/as3/comments/owag5h/building_aab_with_flash_develop_working_experience/

 The thing is: if I include Enhance ane into it, the aab is built well. But if I rename it from .aab to .zip and try to unpack, I get the "zip file corrupted" error.
If I build an .aab without any ane, everything goes well, and renaming it to .zip allows to extract everything

The .zip corruption leads to subsequent errors: in uploading .aab to Enhance or to Google Play. 
I can send all the data needed to fix this.

**Reply by GeneralVimes** [reply]
interesting - I've remade everything from the very beginning, to provide you logs and folders, and now both types of aab - with and without ane, are good, not corrupted zip archives! So, looks like nothing to worry :)

And thank you again for your hard work!

**Reply by ajwfrost** [reply]
@GeneralVimes weird, but good I guess ... if you do find that issue again please let me know. We should be clean-building the AAB file every time you run ADT so it seems a bit odd to be able to create a corrupted zip: if you find this again then please just send us the corrupt zip file and we'll inspect it to see if we can see a problem in the indexing or similar..

thanks

**Reply by marchbold** [reply]
> For the store deployment, presumably you are using the Google Play App Signing? I don't know fully how this works but am wondering whether it's working when the request (or app?) is signed with the key from the play store, but not when it's signed with your 'local' or upload key?

This is highly likely! If you are signing with a different "upload" certificate and testing with that locally Firebase/Google services will likely fail auth, as they use the sha signature of the cert to perform certain actions. 

You could try adding the sha signature fingerprint of your upload cert to the firebase console? It's under "project settings" / "general" / "Your apps".

**Reply by GeneralVimes** [reply]
Something strange happens indeed: the aab shows no errors when unzipping, it is uploaded and accepted to Google Play, but when the game is installed from GP, it crashes: https://github.com/airsdk/Adobe-Runtime-Support/discussions/1039 
Also, there are errors when calling adt -installApp (or, I'm just calling it in the wrong way): https://github.com/airsdk/Adobe-Runtime-Support/discussions/1038

**Reply by stoff99** [reply]
@marchbold @ajwfrost ,

finally i made it. I created the apks without the signing options, in this case bundletool is using the dev.ks file.
I exported the fingerprints of the dev.ks and added it on the android firebase project but this want work.

Now i sign the apks with the keystore file which i generated out of my p12 as i was activating the Google Signing some weeks ago.
Here is my command i use:
`bundletool build-apks --bundle ETeamApp.aab --output output2.apks --connected-device --ks=keystore.ks --ks-pass=pass:YourPassword --ks-key-alias=YourAlias`

If you are not sure about the alias of your ks file use:
`keytool -v -list -keystore keystore.ks`

And than i install the apks:
`bundletool install-apks --apks=output2.apks`

Looks like a mistake from my side because i didn't realize the signing option for the bundle tool .... :(
But hey ... i don't use the CustomResources.ane anymore ... :D
Thx a lot for the help, maybe this will help another one aswell.

**Reply by ajwfrost** [reply]
Great, thanks @stoff99  - and yes, this may well help a lot of other folk too!

**Reply by marchbold** [reply]
Great work @stoff99 !! We need to document this process :) I'm going to start putting together some docs around aab packaging I think...

**Reply by stoff99** [reply]
@marchbold if you need my "raw" documentation about how you can create the needed keystore files to change to Google Play Store Signing, this took me also a while:

[as3-ActivatePlayAppSigningwithp12file-050821-0812-4.pdf](https://github.com/airsdk/Adobe-Runtime-Support/files/6936298/as3-ActivatePlayAppSigningwithp12file-050821-0812-4.pdf)

**Reply by marchbold** [reply]
@stoff99 Brilliant, thank you for that, I'll get it integrated into the air docs site and upload next week.

**Reply by marchbold** [reply]
@stoff99 Just been going through your document and wondering if all those steps are actually needed? Why did you have to convert to PEM and then create the keystore from that. I've had success converting it directly before and just wondering if I'm missing something? 

eg.

`
keytool -importkeystore -srckeystore YOUR_CERTIFICATE.p12 -srcstoretype pkcs12 -srcalias 1 -destkeystore keystore.ks -deststoretype jks -destalias ALIAS
`

Also you don't need this for the upload with pepk, you can use your p12 for this too:

`
java -jar pepk.jar --keystore=YOUR_CERTIFICATE.p12 --alias=1 --output=playappsigning_encrypted_certificate.zip --encryptionkey=XXXXXXXXX
`

Just want to check if you tried this approach and ran into issues or if you just used a bit of a round about approach ;)
